src/libsync/pushnotifications.cpp

```diff
@@ -165,33 +175,128 @@ void PushNotifications::handleAuthenticated()
     qCInfo(lcPushNotifications) << "Authenticated successful on websocket";
     _failedAuthenticationAttemptsCount = 0;
     _isReady = true;
+    startPingTimeoutTimer();
     emit ready();
+
+    // We maybe reconnected to websocket while being offline for a
+    // while. To not miss any notifications that may have happend,
+    // emit all the signals once.
+    emitFilesChanged();
+    emitNotificationsChanged();
+    emitActivitiesChanged();
 }
 
 void PushNotifications::handleNotifyFile()
 {
     qCInfo(lcPushNotifications) << "Files push notification arrived";
-    emit filesChanged(_account);
+    emitFilesChanged();
 }
 
 void PushNotifications::handleInvalidCredentials()
 {
     qCInfo(lcPushNotifications) << "Invalid credentials submitted to websocket";
     if (!tryReconnectToWebSocket()) {
-        _isReady = false;
+        closeWebSocket();
         emit authenticationFailed();
     }
 }
 
 void PushNotifications::handleNotifyNotification()
 {
     qCInfo(lcPushNotifications) << "Push notification arrived";
-    emit notificationsChanged(_account);
+    emitNotificationsChanged();
 }
 
 void PushNotifications::handleNotifyActivity()
 {
     qCInfo(lcPushNotifications) << "Push activity arrived";
+    emitActivitiesChanged();
+}
+
+void PushNotifications::onWebSocketPongReceived(quint64 /*elapsedTime*/, const QByteArray &payload)
+{
+    handleTimeoutPong(payload);
+}
+
+void PushNotifications::handleTimeoutPong(const QByteArray &payload)
+{
+    // We are not interested in different pongs from the server
+    const auto expectedPingPayload = timeoutPingPayload();
+    if (payload != expectedPingPayload) {
+        return;
+    }
+
+    qCDebug(lcPushNotifications) << "Pong received in time";
+
+
+    _timeoutPongReceivedFromWebSocketServer = true;
+    startPingTimeoutTimer();
+}
+
+void PushNotifications::startPingTimeoutTimer()
```

Comment:
@FlexW I'd even propose to get rid of the **Timer** word here as well. Something like _startPingingServer_, _beginPingingServer_, _startSendingPing_ - would be more readable IMHO. The **Timer** part is really not essential here. It's just an implementation detail that is hiding inside this method.

And yes, I agree with @er-vin - **Timeout** is also redundant here.

Reply:
I think the timeout word can add something because it's the ping for checking for timeout. We could also have different kind of pings that check other things. Maybe that's a bit theoretically and we will forever just use one kind of ping but thats why I made the names so specific.